Multiple simultaneous ST instances show error when shut down

I have one (vast) network share on which the variable data is located: the huge HF models and the ST workspace.
To train two LoRAs at the same time I rented two machines that both were using this same network share, so that the HF models don't need to be downloaded multiple times. The training /output showed to two different (sub)directories.

Training worked well.

But when rebooting one machine (so that it could train something else as my setup is not interactive, configured with one training per run for ease of setup) something "strange" happens on the still running, other machine. It shows this error message:
```
[0mProcessing bucket 1.0:  91%|███████████████████████████████████████████████████████████▎     | 52/57 [00:06<00:00,  6.48it/s][0m[0m[0m[RANK 0] [31m2026-03-12 22:53:14,422 [ERROR] Error in queue processing: unable to open database file
Traceback (most recent call last):
  File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/services/cloud/queue/scheduler.py", line 559, in process_loop
    await self.process_queue()
  File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/services/cloud/queue/scheduler.py", line 486, in process_queue
    entry = await self._get_next_ready()
            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/services/cloud/queue/scheduler.py", line 448, in _get_next_ready
    running_count = await self._queue_store.count_running()
                    ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/services/cloud/queue/job_repo_adapter.py", line 75, in count_running
    return await self._job_repo.count_running()
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/services/cloud/storage/job_repository.py", line 801, in count_running
    return await loop.run_in_executor(None, _count)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/services/cloud/storage/job_repository.py", line 793, in _count
    conn = self._get_connection()
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/services/cloud/storage/base.py", line 177, in _get_connection
    conn = sqlite3.connect(
           ^^^^^^^^^^^^^^^^
sqlite3.OperationalError: unable to open database file[0m
[0m
[0m[0m                                                                                                                             [0m
[0m[0m(id=train-1024-image-Julia-image) Bucket 1.0 caching results: {'not_local': 0, 'already_cached': 0, 'cached': 1, 'total': 57}[0m
[0mERROR:    Exception in ASGI application
  + Exception Group Traceback (most recent call last):
  |   File "/opt/venv/lib/python3.12/site-packages/starlette/_utils.py", line 81, in collapse_excgroups
  |     yield
  |   File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 192, in __call__
  |     async with anyio.create_task_group() as task_group:
  |   File "/opt/venv/lib/python3.12/site-packages/anyio/_backends/_asyncio.py", line 783, in __aexit__
  |     raise BaseExceptionGroup(
  | ExceptionGroup: unhandled errors in a TaskGroup (1 sub-exception)
  +-+---------------- 1 ----------------
    | Traceback (most recent call last):
    |   File "/opt/venv/lib/python3.12/site-packages/uvicorn/protocols/http/httptools_impl.py", line 416, in run_asgi
    |     result = await app(  # type: ignore[func-returns-value]
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/opt/venv/lib/python3.12/site-packages/uvicorn/middleware/proxy_headers.py", line 60, in __call__
    |     return await self.app(scope, receive, send)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/opt/venv/lib/python3.12/site-packages/fastapi/applications.py", line 1160, in __call__
    |     await super().__call__(scope, receive, send)
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/applications.py", line 107, in __call__
    |     await self.middleware_stack(scope, receive, send)
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/errors.py", line 186, in __call__
    |     raise exc
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/errors.py", line 164, in __call__
    |     await self.app(scope, receive, _send)
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 191, in __call__
    |     with recv_stream, send_stream, collapse_excgroups():
    |   File "/usr/lib/python3.12/contextlib.py", line 158, in __exit__
    |     self.gen.throw(value)
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/_utils.py", line 87, in collapse_excgroups
    |     raise exc
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 193, in __call__
    |     response = await self.dispatch_func(request, call_next)
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/middleware/security_middleware.py", line 379, in dispatch
    |     return await call_next(request)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 168, in call_next
    |     raise app_exc from app_exc.__cause__ or app_exc.__context__
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 144, in coro
    |     await self.app(scope, receive_or_disconnect, send_no_error)
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 191, in __call__
    |     with recv_stream, send_stream, collapse_excgroups():
    |   File "/usr/lib/python3.12/contextlib.py", line 158, in __exit__
    |     self.gen.throw(value)
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/_utils.py", line 87, in collapse_excgroups
    |     raise exc
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 193, in __call__
    |     response = await self.dispatch_func(request, call_next)
    |                ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/middleware/security_middleware.py", line 112, in dispatch
    |     response = await call_next(request)
    |                ^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 168, in call_next
    |     raise app_exc from app_exc.__cause__ or app_exc.__context__
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 144, in coro
    |     await self.app(scope, receive_or_disconnect, send_no_error)
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/cors.py", line 87, in __call__
    |     await self.app(scope, receive, send)
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/exceptions.py", line 63, in __call__
    |     await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    |     raise exc
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/_exception_handler.py", line 42, in wrapped_app
    |     await app(scope, receive, sender)
    |   File "/opt/venv/lib/python3.12/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
    |     await self.app(scope, receive, send)
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/routing.py", line 716, in __call__
    |     await self.middleware_stack(scope, receive, send)
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/routing.py", line 736, in app
    |     await route.handle(scope, receive, send)
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/routing.py", line 290, in handle
    |     await self.app(scope, receive, send)
    |   File "/opt/venv/lib/python3.12/site-packages/fastapi/routing.py", line 130, in app
    |     await wrap_app_handling_exceptions(app, request)(scope, receive, send)
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    |     raise exc
    |   File "/opt/venv/lib/python3.12/site-packages/starlette/_exception_handler.py", line 42, in wrapped_app
    |     await app(scope, receive, sender)
    |   File "/opt/venv/lib/python3.12/site-packages/fastapi/routing.py", line 116, in app
    |     response = await f(request)
    |                ^^^^^^^^^^^^^^^^
    |   File "/opt/venv/lib/python3.12/site-packages/fastapi/routing.py", line 453, in app
    |     solved_result = await solve_dependencies(
    |                     ^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/opt/venv/lib/python3.12/site-packages/fastapi/dependencies/utils.py", line 675, in solve_dependencies
    |     solved = await call(**solved_result.values)
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/services/cloud/auth/middleware.py", line 410, in get_current_user
    |     auth = await get_auth_context(request)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/services/cloud/auth/middleware.py", line 397, in get_auth_context
    |     return await middleware(request)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/services/cloud/auth/middleware.py", line 267, in __call__
    |     result = await self.store.authenticate_api_key(api_key)
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/services/cloud/auth/user_store.py", line 416, in authenticate_api_key
    |     result = await self._api_keys.authenticate(raw_key)
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/services/cloud/auth/stores/api_key_store.py", line 178, in authenticate
    |     return await loop.run_in_executor(None, _auth)
    |            ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/usr/lib/python3.12/concurrent/futures/thread.py", line 58, in run
    |     result = self.fn(*self.args, **self.kwargs)
    |              ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
    |   File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/services/cloud/auth/stores/api_key_store.py", line 139, in _auth
    |     conn = self._get_connection()
    |            ^^^^^^^^^^^^^^^^^^^^^^
    |   File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/services/cloud/auth/stores/base.py", line 86, in _get_connection
    |     conn = sqlite3.connect(
    |            ^^^^^^^^^^^^^^^^
    | sqlite3.OperationalError: unable to open database file
    +------------------------------------

The above exception was the direct cause of the following exception:

Traceback (most recent call last):
  File "/opt/venv/lib/python3.12/site-packages/uvicorn/protocols/http/httptools_impl.py", line 416, in run_asgi
    result = await app(  # type: ignore[func-returns-value]
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.12/site-packages/uvicorn/middleware/proxy_headers.py", line 60, in __call__
    return await self.app(scope, receive, send)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.12/site-packages/fastapi/applications.py", line 1160, in __call__
    await super().__call__(scope, receive, send)
  File "/opt/venv/lib/python3.12/site-packages/starlette/applications.py", line 107, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/errors.py", line 186, in __call__
    raise exc
  File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/errors.py", line 164, in __call__
    await self.app(scope, receive, _send)
  File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 191, in __call__
    with recv_stream, send_stream, collapse_excgroups():
  File "/usr/lib/python3.12/contextlib.py", line 158, in __exit__
    self.gen.throw(value)
  File "/opt/venv/lib/python3.12/site-packages/starlette/_utils.py", line 87, in collapse_excgroups
    raise exc
  File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 193, in __call__
    response = await self.dispatch_func(request, call_next)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/middleware/security_middleware.py", line 379, in dispatch
    return await call_next(request)
           ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 168, in call_next
    raise app_exc from app_exc.__cause__ or app_exc.__context__
  File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 144, in coro
    await self.app(scope, receive_or_disconnect, send_no_error)
  File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 191, in __call__
    with recv_stream, send_stream, collapse_excgroups():
  File "/usr/lib/python3.12/contextlib.py", line 158, in __exit__
    self.gen.throw(value)
  File "/opt/venv/lib/python3.12/site-packages/starlette/_utils.py", line 87, in collapse_excgroups
    raise exc
  File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 193, in __call__
    response = await self.dispatch_func(request, call_next)
               ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/middleware/security_middleware.py", line 112, in dispatch
    response = await call_next(request)
               ^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 168, in call_next
    raise app_exc from app_exc.__cause__ or app_exc.__context__
  File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/base.py", line 144, in coro
    await self.app(scope, receive_or_disconnect, send_no_error)
  File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/cors.py", line 87, in __call__
    await self.app(scope, receive, send)
  File "/opt/venv/lib/python3.12/site-packages/starlette/middleware/exceptions.py", line 63, in __call__
    await wrap_app_handling_exceptions(self.app, conn)(scope, receive, send)
  File "/opt/venv/lib/python3.12/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    raise exc
  File "/opt/venv/lib/python3.12/site-packages/starlette/_exception_handler.py", line 42, in wrapped_app
    await app(scope, receive, sender)
  File "/opt/venv/lib/python3.12/site-packages/fastapi/middleware/asyncexitstack.py", line 18, in __call__
    await self.app(scope, receive, send)
  File "/opt/venv/lib/python3.12/site-packages/starlette/routing.py", line 716, in __call__
    await self.middleware_stack(scope, receive, send)
  File "/opt/venv/lib/python3.12/site-packages/starlette/routing.py", line 736, in app
    await route.handle(scope, receive, send)
  File "/opt/venv/lib/python3.12/site-packages/starlette/routing.py", line 290, in handle
    await self.app(scope, receive, send)
  File "/opt/venv/lib/python3.12/site-packages/fastapi/routing.py", line 130, in app
    await wrap_app_handling_exceptions(app, request)(scope, receive, send)
  File "/opt/venv/lib/python3.12/site-packages/starlette/_exception_handler.py", line 53, in wrapped_app
    raise exc
  File "/opt/venv/lib/python3.12/site-packages/starlette/_exception_handler.py", line 42, in wrapped_app
    await app(scope, receive, sender)
  File "/opt/venv/lib/python3.12/site-packages/fastapi/routing.py", line 116, in app
    response = await f(request)
               ^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.12/site-packages/fastapi/routing.py", line 453, in app
    solved_result = await solve_dependencies(
                    ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/opt/venv/lib/python3.12/site-packages/fastapi/dependencies/utils.py", line 675, in solve_dependencies
    solved = await call(**solved_result.values)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/services/cloud/auth/middleware.py", line 410, in get_current_user
    auth = await get_auth_context(request)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/services/cloud/auth/middleware.py", line 397, in get_auth_context
    return await middleware(request)
           ^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/services/cloud/auth/middleware.py", line 267, in __call__
    result = await self.store.authenticate_api_key(api_key)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/services/cloud/auth/user_store.py", line 416, in authenticate_api_key
    result = await self._api_keys.authenticate(raw_key)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/services/cloud/auth/stores/api_key_store.py", line 178, in authenticate
    return await loop.run_in_executor(None, _auth)
           ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/usr/lib/python3.12/concurrent/futures/thread.py", line 58, in run
    result = self.fn(*self.args, **self.kwargs)
             ^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^^
  File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/services/cloud/auth/stores/api_key_store.py", line 139, in _auth
    conn = self._get_connection()
           ^^^^^^^^^^^^^^^^^^^^^^
  File "/app/SimpleTuner/simpletuner/simpletuner_sdk/server/services/cloud/auth/stores/base.py", line 86, in _get_connection
    conn = sqlite3.connect(
           ^^^^^^^^^^^^^^^^
sqlite3.OperationalError: unable to open database file
[0m[0m[0m[RANK 0] [33m2026-03-12 22:53:17,016 [WARNING] Could not send webhook request to http://0.0.0.0:8001/callback: 500 Server Error: Internal Server Error for url: http://0.0.0.0:8001/callback[0m
```

Expected behavior: shutting down one machine / running instance shouldn't affect the other.

Possible work around when that is not possible: have a command line parameter that tells the ST instance that it is in a shared environment or one telling it which number it is and so it creates all of its possibly shared and conflicting data (only the SQLite DB?) with this index in the file name, so that multiple instances can coexist without interfering each other.

--- a/simpletuner/cli/server.py
+++ b/simpletuner/cli/server.py
@@ -118,6 +118,16 @@ def _setup_ssl_config(ssl_key: Optional[str] = None, ssl_cert: Optional[str] = N
         return None
 
 
+def _ensure_server_state_dir() -> None:
+    """Keep server SQLite state local unless the user explicitly overrides it."""
+    if os.environ.get("SIMPLETUNER_STATE_DIR"):
+        return
+
+    from simpletuner.simpletuner_sdk.server.services.cloud.storage.base import get_local_state_dir
+
+    os.environ["SIMPLETUNER_STATE_DIR"] = str(get_local_state_dir())
+
+
 def cmd_server(args) -> int:
     """Handle server command."""
     host = getattr(args, "host", "0.0.0.0")
@@ -129,6 +139,8 @@ def cmd_server(args) -> int:
     ssl_cert = getattr(args, "ssl_cert", None)
     ssl_no_verify = getattr(args, "ssl_no_verify", False)
     env = getattr(args, "env", None)
+
+    _ensure_server_state_dir()
 
     if not os.environ.get("SIMPLETUNER_CONFIG_DIR"):
         webui_configs_dir = _get_webui_configs_dir()

--- a/simpletuner/service_worker.py
+++ b/simpletuner/service_worker.py
@@ -11,6 +11,11 @@ from fastapi.responses import JSONResponse
 from fastapi.staticfiles import StaticFiles
 from pydantic import BaseModel
 
+from simpletuner.simpletuner_sdk.server.services.cloud.storage.base import get_local_state_dir
+
+os.environ.setdefault("SIMPLETUNER_STATE_DIR", str(get_local_state_dir()))
+os.environ.setdefault("SIMPLETUNER_SERVER_ROOT_PID", str(os.getpid()))
+
 from simpletuner.simpletuner_sdk.configuration import Configuration
 
 # Import route modules
@@ -20,8 +25,6 @@ from simpletuner.simpletuner_sdk.server.routes.publishing import router as publi
 from simpletuner.simpletuner_sdk.server.routes.web import router as web_router
 from simpletuner.simpletuner_sdk.server.utils.paths import get_config_directory, get_static_directory, get_template_directory
 from simpletuner.simpletuner_sdk.training_host import TrainingHost
-
-os.environ.setdefault("SIMPLETUNER_SERVER_ROOT_PID", str(os.getpid()))
 
 
 # Pydantic models for request/response

--- a/simpletuner/simpletuner_sdk/server/services/cloud/async_job_store.py
+++ b/simpletuner/simpletuner_sdk/server/services/cloud/async_job_store.py
@@ -21,6 +21,7 @@ from __future__ import annotations
 
 import asyncio
 import logging
+import os
 from pathlib import Path
 from typing import Any, Dict, List, Optional
 
@@ -104,12 +105,16 @@ class AsyncJobStore:
             return cls._instance
 
     def _resolve_config_dir(self) -> Path:
-        """Resolve config directory from WebUIStateStore or default.
+        """Resolve the server state directory.
 
-        Returns the SimpleTuner root directory (e.g. /notebooks/simpletuner
-        or ~/.simpletuner). The database path is then constructed as
-        config_dir / "cloud" / "jobs.db".
+        Returns the SimpleTuner state root directory. The database path is then
+        constructed as state_dir / "cloud" / "jobs.db".
         """
+        from .storage.base import get_default_config_dir
+
+        if os.environ.get("SIMPLETUNER_STATE_DIR"):
+            return get_default_config_dir()
+
         try:
             from .webui_state import WebUIStateStore
 
@@ -119,8 +124,6 @@ class AsyncJobStore:
                 return Path(defaults.configs_dir)
             return store.base_dir.parent
         except Exception:
-            from .storage.base import get_default_config_dir
-
             return get_default_config_dir()
 
     async def _ensure_initialized(self) -> None:

--- a/simpletuner/simpletuner_sdk/server/services/cloud/storage/base.py
+++ b/simpletuner/simpletuner_sdk/server/services/cloud/storage/base.py
@@ -21,22 +21,42 @@ logger = logging.getLogger(__name__)
 T = TypeVar("T")
 
 
-def get_default_config_dir() -> Path:
-    """Get the default configuration directory in a cross-platform manner.
+def get_local_state_dir() -> Path:
+    """Get the local per-machine SimpleTuner state directory.
 
-    Environment variable override:
-        SIMPLETUNER_CONFIG_DIR: If set, use this path directly (useful for testing)
+    This intentionally avoids shared ML workspace paths. SQLite-backed server
+    state should stay local to the machine unless explicitly configured.
+    """
+    if sys.platform == "win32":
+        base = Path(os.environ.get("APPDATA", Path.home() / "AppData" / "Roaming"))
+        return base / "SimpleTuner"
+    if sys.platform == "darwin":
+        return Path.home() / "Library" / "Application Support" / "SimpleTuner"
+    return Path.home() / ".simpletuner"
+
+
+def get_default_config_dir() -> Path:
+    """Get the default server state directory in a cross-platform manner.
+
+    Environment variable overrides:
+        SIMPLETUNER_STATE_DIR: If set, use this path for server state.
+        SIMPLETUNER_CONFIG_DIR: Legacy fallback for tests/custom deployments.
 
     Returns:
-        Path to the configuration directory:
+        Path to the server state directory:
         - Windows: %APPDATA%/SimpleTuner
         - macOS: ~/Library/Application Support/SimpleTuner
         - Linux: /workspace/simpletuner, /notebooks/simpletuner, or ~/.simpletuner
     """
-    # Check for explicit override (testing, custom deployment)
+    state_override = os.environ.get("SIMPLETUNER_STATE_DIR")
+    if state_override:
+        return Path(state_override).expanduser()
+
+    # Legacy override retained for tests and deployments that already use it
+    # as a state root. New server entry points set SIMPLETUNER_STATE_DIR.
     override = os.environ.get("SIMPLETUNER_CONFIG_DIR")
     if override:
-        return Path(override)
+        return Path(override).expanduser()
 
     if sys.platform == "win32":
         base = Path(os.environ.get("APPDATA", Path.home() / "AppData" / "Roaming"))

--- a/simpletuner/simpletuner_sdk/server/webui_app.py
+++ b/simpletuner/simpletuner_sdk/server/webui_app.py
@@ -1,6 +1,11 @@
 """Create app instance for running with uvicorn."""
 
+import os
+
 from .app import create_unified_app
+from .services.cloud.storage.base import get_local_state_dir
+
+os.environ.setdefault("SIMPLETUNER_STATE_DIR", str(get_local_state_dir()))
 
 # Create the app instance
 app = create_unified_app()

--- a/tests/test_cloud_state_paths.py
+++ b/tests/test_cloud_state_paths.py
@@ -1,0 +1,66 @@
+import json
+import os
+import tempfile
+import unittest
+from pathlib import Path
+
+from simpletuner.cli.server import _ensure_server_state_dir
+from simpletuner.simpletuner_sdk.server.services.cloud.async_job_store import AsyncJobStore
+from simpletuner.simpletuner_sdk.server.services.cloud.storage.base import get_default_config_dir, get_default_db_path
+
+
+class TestCloudStatePaths(unittest.TestCase):
+    def setUp(self) -> None:
+        self._tmpdir = tempfile.TemporaryDirectory()
+        self.tmp_path = Path(self._tmpdir.name)
+        self._previous = {
+            "SIMPLETUNER_STATE_DIR": os.environ.get("SIMPLETUNER_STATE_DIR"),
+            "SIMPLETUNER_CONFIG_DIR": os.environ.get("SIMPLETUNER_CONFIG_DIR"),
+            "SIMPLETUNER_WEB_UI_CONFIG": os.environ.get("SIMPLETUNER_WEB_UI_CONFIG"),
+        }
+
+    def tearDown(self) -> None:
+        for key, value in self._previous.items():
+            if value is None:
+                os.environ.pop(key, None)
+            else:
+                os.environ[key] = value
+        AsyncJobStore._instance = None
+        self._tmpdir.cleanup()
+
+    def test_state_dir_takes_precedence_over_config_dir(self) -> None:
+        state_dir = self.tmp_path / "local-state"
+        config_dir = self.tmp_path / "shared-config"
+        os.environ["SIMPLETUNER_STATE_DIR"] = str(state_dir)
+        os.environ["SIMPLETUNER_CONFIG_DIR"] = str(config_dir)
+
+        self.assertEqual(get_default_config_dir(), state_dir)
+        self.assertEqual(get_default_db_path("jobs.db"), state_dir / "cloud" / "jobs.db")
+
+    def test_async_job_store_uses_state_dir_before_webui_configs_dir(self) -> None:
+        state_dir = self.tmp_path / "local-state"
+        shared_configs = self.tmp_path / "shared" / "config"
+        webui_dir = self.tmp_path / "webui"
+        webui_dir.mkdir()
+        shared_configs.mkdir(parents=True)
+        (webui_dir / "defaults.json").write_text(
+            json.dumps({"configs_dir": str(shared_configs)}),
+            encoding="utf-8",
+        )
+        os.environ["SIMPLETUNER_STATE_DIR"] = str(state_dir)
+        os.environ["SIMPLETUNER_WEB_UI_CONFIG"] = str(webui_dir)
+
+        store = AsyncJobStore()
+
+        self.assertEqual(store.get_database_path(), state_dir / "cloud" / "jobs.db")
+
+    def test_server_sets_local_state_dir_when_unset(self) -> None:
+        os.environ.pop("SIMPLETUNER_STATE_DIR", None)
+
+        _ensure_server_state_dir()
+
+        self.assertEqual(Path(os.environ["SIMPLETUNER_STATE_DIR"]).expanduser(), get_default_config_dir())
+
+
+if __name__ == "__main__":
+    unittest.main()

